issue #3 solution

RomanZorkin opened the pull request:
Implement /share API endpoint and output the URL of a file on S3 #3

RomanZorkin committed:
issue #28 solution

Create an integration test for /share endpoint #28

RomanZorkin committed:
Update poetry dependency

Update poetry dependency

RomanZorkin committed:
Poetry add boto3

Update dependency

anatoly-scherbakov commented on geopatterns_demo/file_handler.py:
```diff
@@ -0,0 +1,74 @@
+
+
+def file_to_s3(file_name: str, image: BytesIO) -> str:
+    """Save file to S3 bucket & return it's URL.
```
```suggestion
    """Save file to S3 bucket & return its URL.
```

anatoly-scherbakov commented on geopatterns_demo/file_handler.py:
```diff
@@ -0,0 +1,74 @@
+    Returns:
+        bool: bool
+    """
+    bucket_object = list(bucket.objects.filter(Prefix=file_name))
```
This method of finding an object on a bucket is probably not optimal. A proper method would be using, say, this: https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Client.head_object

But in this particular case I believe we should not do it at all. See below.

anatoly-scherbakov commented on geopatterns_demo/file_handler.py:
```diff
@@ -0,0 +1,74 @@
+    s3 = boto3.Resource('s3')
+    bucket = s3.Bucket(settings.S3_BUCKET_NAME)
+    bucket.upload_fileobj(image.read(), file_name)
+    if check_file(file_name, bucket):
```
This probably will not work.

S3 is a storage that is eventually consistent. Eventual consistency means that, once you've uploaded a file to S3, it will appear there and will be available for retrieval but that is not guaranteed to happen momentarily.

There might be a delay — I do not know how long but, say, 0.5 sec — before the file appears. So, this `check_file()` call can easily miss the file because it is executed without a delay.

Do we actually **need** to check the existence of the uploaded file here?

I think not. S3 is a well tested system. We just need to upload the file and send the URL to the user, that is it.

anatoly-scherbakov reviewed (changes requested):
Please see comments in the PR. The code can be much simplified, I believe. I will hopefully look into integrating tests into the project shortly.

RomanZorkin committed:
Adjustments have been made according to the comments

Changes have been made in accordance with https://github.com/anatoly-scherbakov/geopatterns-demo/pull/32#pullrequestreview-797642871

RomanZorkin committed:
Update poetry project

Update poetry project

RomanZorkin commented:
Adjustments have been made according to the comments 

Changes have been made in accordance with #32 (review)

RomanZorkin commented:
Adjustments have been made according to the comments
Changes have been made in accordance with #32 (review)

RomanZorkin committed:
Fixed some bugs according to linter rules. Several dependencies have been updated.
